Import dashboard from http url or grafana.com dashboard ID

seuf opened the pull request:
##### SUMMARY
Update the grafana dashboard module to allow HTTP url or Grafana dashboard ID when importing.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/monitoring/grafana_dashboard.py

##### ADDITIONAL INFORMATION
The `path` parameter of the Grafana dashboard module now accept a HTTP url or a public grafana.com dashboard ID.
There is also 2 aliases for this parameter. Example :
```yaml
    - name: Import Grafana dashboard Zabbix
      grafana_dashboard:
        grafana_url: http://grafana.company.com
        grafana_api_key: "{{ grafana_api_key }}"
        folder: zabbix
        dashboard_id: 6098

    - name: Import Grafana dashboard zabbix
      grafana_dashboard:
        grafana_url: http://grafana.company.com
        grafana_api_key: "{{ grafana_api_key }}"
        folder: zabbix
        dashboard_url: https://grafana.com/api/dashboards/6098/revisions/1/download
```

Related Issues / pulls : 
https://github.com/ansible/ansible/issues/52636
https://github.com/ansible/ansible/pull/61716

seuf committed:
[grafana_dashboard] Allow dashboard url or dashboard id from grafana.com when importing dashboard

rrey commented on lib/ansible/modules/monitoring/grafana_dashboard.py:
```diff
@@ -298,11 +314,20 @@ def grafana_dashboard_changed(payload, dashboard):
+            with open(data['path'], 'r') as json_file:
+                payload = json.load(json_file)
+        except Exception as e:
+            raise GrafanaAPIException("Can't load json file %s" % to_native(e))
```
I think you should use new parameters instead of creating aliases on `path`.
It will allow you to properly document the usage of each parameters.

In the current implementation, you cannot provide the revision of the dashboard if you provide an id as the revision is hardcoded to "1" in the url. So you need an additional parameter that specifies the revision.

I guess that `path` could be a local path on disk or an url.
But it should not accept an id for a location on grafana.com. This case should have dedicated parameters for the dashboard id and revision.

seuf replied:
OK, fixed :)

till commented:
Merci, @seuf! :)

rrey commented on lib/ansible/modules/monitoring/grafana_dashboard.py:
```diff
@@ -141,7 +167,7 @@
 
 import json
-
+import re
```
I think it is now unused, right ?

seuf replied:
right. with the dedicated `dashboard_id` no need to regex match the path to check if its an id.

rrey reviewed (changes requested):
Would you mind adding an integration test ?

seuf committed:
added integration tests to grafana dashboard module

seuf commented:
Hello @rrey ! I've added some integration tests :

- import from url
- import from id
- import from file
- delete dashboard

Also I've fixed a small bug i discovered thanks to integration tests : we need to remove `id` field from payload when present.

rrey commented:
LGTM

rrey commented:
integration tests FTW ! 😁

seuf commented:
rebased because grafana module are moved to grafana module directory.

seuf commented:
shipit

rrey commented:
shipit

gundalow commented:
Replaced by Collection PR: https://github.com/gundalow-collections/grafana/pull/11

anilpanditi commented:
---
- name: task to parameterise grafana template
  hosts: localhost
  become: true
  user: root
  tasks:
    - name: Import Grafana dashboard
      grafana_dashboard:
        grafana_url: http://<host-ip>:3000
        state: present
        message: Updated by ansible
        uid: edO3sTlipu
        overwrite: yes
        path: "/home/centos/Kubernetes_cluster_monitoring.json"

And i am getting error below.
"msg": "error : Unable to create the new dashboard edO3sTlipu : 404 - {'status': 404, 'body': '{\"message\":\"Dashboard not found\",\"status\":\"not-found\"}', 'content-length': '54', 'url': 'http://13.235.2.156:3000/api/dashboards/db', 'expires': '-1', 'msg': 'HTTP Error 404: Not Found', 'connection': 'close', 'pragma': 'no-cache', 'cache-control': 'no-cache', 'date': 'Tue, 07 Jan 2020 08:38:46 GMT', 'x-frame-options': 'deny', 'content-type': 'application/json'}."

rrey commented:
Hi @anilpanditi,

Your issue us unrelated to this (Closed) PR.
Please create an issue dedicated to your issue and provide the minimum information required (Ansible version and Grafana version).
